provider/spark.go

```diff
@@ -903,7 +926,106 @@ func (spark *SparkOfflineStore) DeleteMaterialization(id MaterializationID) erro
 	return nil
 }
 
-func (spark *SparkOfflineStore) CreateTrainingSet(TrainingSetDef) error {
+type S3TrainingSet struct {
+	id       ResourceID
+	store    SparkStore
+	Key      string
+	err      error
+	label    interface{}
+	features []interface{}
+	iter     chan []byte
+	rows     int64
+	idx      int64
+}
+
+func trainingSetValuesFromCSV(csv string) ([]interface{}, interface{}) {
+	values := strings.Split(string(csv), ",")
+	numFeatures := len(values) - 1
+	features := make([]interface{}, numFeatures)
+	for i := 0; i < numFeatures; i++ {
+		features[i] = reflect.ValueOf(values[i]).Interface()
+	}
+	lastIndex := numFeatures
+	label := reflect.ValueOf(values[lastIndex]).Interface()
+	return features, label
+}
+
+func (s *S3TrainingSet) Next() bool {
+	if s.idx >= s.rows {
+		return false
+	}
+	if s.iter == nil {
+		iterator, err := s.store.ResourceStream(s.Key)
+		if err != nil {
+			s.err = err
+			return false
+		}
+		s.iter = iterator
+	}
+	val := <-s.iter
+	features, label := trainingSetValuesFromCSV(string(val))
+	s.features = features
+	s.label = label
+	s.idx += 1
+	return true
+}
+
+func (s *S3TrainingSet) Features() []interface{} {
+	return s.features
+}
+
+func (s *S3TrainingSet) Label() interface{} {
+	return s.label
+}
+
+func (s *S3TrainingSet) Err() error {
+	return s.err
+}
+
+func (spark *SparkOfflineStore) registeredResourceSchema(id ResourceID) (ResourceSchema, error) {
+	table, err := spark.GetResourceTable(id)
+	if err != nil {
+		return ResourceSchema{}, fmt.Errorf("resource not registered: %v", err)
+	}
+	sparkResourceTable, ok := table.(*S3OfflineTable)
+	if !ok {
+		return ResourceSchema{}, fmt.Errorf("could not convert offline table with id %v to sparkResourceTable", id)
+	}
+	return sparkResourceTable.schema, nil
+}
+
+func (spark *SparkOfflineStore) CreateTrainingSet(def TrainingSetDef) error {
+	sourcePaths := make([]string, 0)
+	featureSchemas := make([]ResourceSchema, 0)
+	destinationPath := spark.Store.ResourcePath(def.ID)
+	exists, _ := spark.Store.FileExists(destinationPath)
+	if exists {
+		return fmt.Errorf("training set already exists %v already exists", def.ID)
+	}
+	labelSchema, err := spark.registeredResourceSchema(def.Label)
+	if err != nil {
+		return fmt.Errorf("Could not get schema of label %s: %v", def.Label, err)
+	}
+	labelPath := fmt.Sprintf("%s%s", spark.Store.BucketPrefix(), labelSchema.SourceTable)
```

Comment:
Is this reused? Should it be a function?

Reply:
changed it to a function